src/test/java/com/radixdlt/crypto/SignaturesTest.java

```diff
@@ -21,14 +26,85 @@ public void verify_that_default_signature_scheme_is_ecdsa() {
         assertEquals(nonEmptySignatures.keyToSignatures().size(), 1);
     }
 
+    @Test(expected = Test.None.class /* no exception expected */)
+    public void test_that_we_can_bulk_verify_signatures_verify_all_two() throws CryptoException {
+        test_that_we_can_bulk_verify_signatures(
+                2,
+                true,
+                2,
+                0
+        );
+    }
+
+    @Test(expected = Test.None.class /* no exception expected */)
+    public void test_that_we_can_bulk_verify_signatures_verify_at_least_N_of_M() throws CryptoException {
+        for (int valid = 0; valid < 5; valid++) {
+            for (int invalid = 0; invalid < 5; invalid++) {
+                for (int threshold = 0; threshold <= (valid + invalid); threshold++) {
+
+                    boolean isExpectedToMeetThreshold = valid >= threshold;
+                    test_that_we_can_bulk_verify_signatures(
+                            threshold,
+                            isExpectedToMeetThreshold,
+                            valid,
+                            invalid
+                    );
+                }
+
+            }
+        }
+    }
+
+    private void test_that_we_can_bulk_verify_signatures(
+            int thresholdNumberOfValidSignatures,
+            boolean isExpectedToMeetThreshold,
+            int numberOfValidSignaturesToCreate,
+            int numberOfInvalidSignaturesToCreate
+    ) throws CryptoException {
+        if (thresholdNumberOfValidSignatures > (numberOfValidSignaturesToCreate + numberOfInvalidSignaturesToCreate)) {
+            throw new IllegalArgumentException(
+                    "The sum of valid + invalid signatures should >= 'thresholdNumberOfValidSignatures'"
+            );
+        }
+        Signatures signatures = DefaultSignatures.emptySignatures();
+        Hash hashedMessage = new Hash(Hash.hash256("You must do what you feel is right of course".getBytes()));
+        for (int i = 0; i < numberOfValidSignaturesToCreate + numberOfInvalidSignaturesToCreate; i++) {
+            ECKeyPair keyPair = new ECKeyPair();
+            assertNotNull(keyPair);
+            Signature signature = null;
+            boolean shouldSignatureBeValid = i >= numberOfInvalidSignaturesToCreate;
+            if (shouldSignatureBeValid) {
+                signature = keyPair.sign(hashedMessage);
+            } else {
+                signature = new ECDSASignature(BigInteger.ONE, BigInteger.ONE);
+            }
+            assertNotNull(signature);
+
+            signatures = signatures.concatenate(keyPair.getPublicKey(), signature);
+        }
+
+        assertEquals((numberOfInvalidSignaturesToCreate + numberOfValidSignaturesToCreate), signatures.keyToSignatures().size());
+        boolean doesSignatureMeetValidityThreshold = signatures.hasSignedMessage(hashedMessage, thresholdNumberOfValidSignatures);
```

Comment:
I think this is more complicated than it needs to be. It would be more expressive as separate assertions for both cases of isExpectedToMeetThreshold, with proper assertion messages.

Reply:
Please be verbose and write your proposal verbatim 😄  (since I'm new to this...)

Do you mean something like:

```java
assertEquals((numberOfInvalidSignaturesToCreate + numberOfValidSignaturesToCreate), signatures.keyToSignatures().size());
        assertEquals(String.format("Expected signatureThreshold result to be: '%s', but it was the inverse.", isExpectedToMeetThreshold), isExpectedToMeetThreshold, signatures.hasSignedMessage(hashedMessage, thresholdNumberOfValidSignatures));
    }
```

?

Since not having the variable `doesSignatureMeetValidityThreshold`, we dont really wanna call  the function `hasSignedMessage` twice, so I can't really display anything else than ", but it was the inverse.". I can of course do:

```java
String.format(
    "Expected signatureThreshold result to be: '%s', but it was the inverse.", 
    isExpectedToMeetThreshold,
    !isExpectedToMeetThreshold
)
```

Notice the `!isExpectedToMeetThreshold`, but that is kind of confusing to read in code, even though it would produce the proper assertion fail message. But what other "inverse" couldd a `boolean` have?

Reply:
@flotothemoon ping ^